### PR TITLE
feat: File normalization function

### PR DESCRIPTION
**Pull request by edas**
As requested in https://github.com/cozy/cozy-notes/pull/41#discussion_r360844792

### CODE REVIEW

File: packages/cozy-client/src/models/file.js [hunk 1]
```diff
@@ -30,3 +30,27 @@ export const isNote = file => {
...
+  const id = file._id || file.id
+  const type = file.type || file._type || 'io.cozy.files'
+  const path = parent ? parent.path + '/' + file.name : undefined
+  return {
+    _id: id,
+    id: id,
+    _type: type,
+    type: type,
+    path: path,
+    ...(file || {})
```
Comment (drazik) [inline packages/cozy-client/src/models/file.js]: Nit: `...file` should be enough.

```js
const obj = { ...undefined }
console.log(obj)
// {}
```

File: packages/cozy-client/src/models/file.spec.js [hunk 2]
```diff
@@ -29,4 +29,36 @@ describe('File Model', () => {
...
+      const doc = { id }
+      expect(file.normalize(doc)._type).toBe('io.cozy.files')
+    })
+
+    it('should define the path if parent is provided', () => {
+      const doc = { id, name: 'myfile.ext' }
+      const parent = { path: '/my/path' }
+      expect(file.normalize(doc, parent).path).toBe('/my/path/myfile.ext')
+    })
+  })
```
Comment (Crash--) [inline packages/cozy-client/src/models/file.spec.js]: Can you add a test when there is no path? Since you have a condition on it

Comment (edas) [reply packages/cozy-client/src/models/file.spec.js]: A parent is a directory, so it should always have a path. I may throw an error if a path is missing but I usualy test only how a function should be used, and not what erroneous parameters could do.

Comment (edas) [reply packages/cozy-client/src/models/file.spec.js]: I have added two tests, one on missing `parent` and one for a `parent` without `path`. Is it what you missed ?

File: packages/cozy-client/src/models/file.js [hunk 3]
```diff
@@ -30,3 +30,27 @@ export const isNote = file => {
     return true
   return false
 }
+
+/**
+ * Normalize an object representing a io.cozy.file object
+ *
+ * Ensure existence of `_id` and `_type` and define a `path`
```
Comment (ptbrowne) [inline packages/cozy-client/src/models/file.js]:
```suggestion
 * Ensures existence of `_id` and `_type` and defines a `path`
```

File: packages/cozy-client/src/models/file.js [hunk 4]
```diff
@@ -30,3 +30,27 @@ export const isNote = file => {
...
+ * on files if provided the parent directory.
```
Comment (ptbrowne) [inline packages/cozy-client/src/models/file.js]:
```suggestion
 * on file if the parent directory is provided.
```

File: packages/cozy-client/src/models/file.js [hunk 5]
```diff
@@ -30,3 +30,27 @@ export const isNote = file => {
...
+ *
+ * @param {object} file - object representing the file
+ * @param {object} parent - object representing the parent directory
+ * @returns {object} full normalized object
```
Comment (ptbrowne) [inline packages/cozy-client/src/models/file.js]: Is it a private or public method ?

File: packages/cozy-client/src/models/file.js [hunk 6]
```diff
@@ -30,3 +30,27 @@ export const isNote = file => {
...
+  const type = file.type || file._type || 'io.cozy.files'
+  const path = parent && parent.path ? parent.path + '/' + file.name : undefined
+  return {
+    _id: id,
+    id: id,
+    _type: type,
+    type: type,
+    path: path,
+    ...file
+  }
```
Comment (ptbrowne) [inline packages/cozy-client/src/models/file.js]: In other parts of cozy-client's code, normalize only deals with _id and _type. Here we add logic about the path, are we sure that it should be done in another function ?

Also, what happens if file has a path and parent has a path ? Here the file's path would override the parent's path, is is something that we want ? It seems like an unwanted side-effect, this is why I am a bit hesitant with having the path logic inside the normalize method.

Comment (edas) [reply packages/cozy-client/src/models/file.js]: Logic was "Ensure it has a path, build it from parent if needed", like we ensure an _id and a _type (but won't override the document's ones if they are present)

File: packages/cozy-client/src/models/file.js (hunk 5 above)
Comment (edas) [reply packages/cozy-client/src/models/file.js]: Today it's used by cozy-sharing, so public.

I've added the corresponding jsdoc

**Comment by edas** [comment]
I've separated the two methods, let me know if it's better for you

**Pull request merged**